### ISSUE TITLE
Error: Failed to build headers.

### ISSUE THREAD

**Issue by tonymathew17sap** [SAP/cloud-sdk-js#4033]
**Describe the bug**
Getting the error "Error: Failed to build headers." when connecting with destination of authentication type "OAuth2SAMLBearerAssertion". I'm using open-api client 'execute()' method with 'addCustomHeaders()' method to add headers

**To Reproduce**

      const response = await OpenApi.Client('id')
        .addCustomHeaders({
          ['custom-header']: `dummy-value`,
        })
        .execute({
          destinationName: 'test',
        });

This is giving me the error "Error: Failed to build headers."

**Used Versions:**

- node version via `node -v` v18.16.0
- npm version via `npm -v` 9.7.1
- SAP Cloud SDK version you used as dependency:     "@sap-cloud-sdk/openapi": "^3.4.0",

**Impact / Priority**

Affected development phase: Production

Impact: Blocked

**Comment by tomfrenken** [comment]
Hey @tonymathew17sap,

For destinations of type `OAuth2SAMLBearerAssertion` you need to provide a user jwt, otherwise, building the headers will fail. You may want to check out [our documentation](https://sap.github.io/cloud-sdk/docs/js/features/connectivity/destinations#authentication-and-json-web-token-jwt-retrieval) regarding the different auth types and their associated auth flows.

In your case, you need to provide the jwt in the execute method, here is an example:
```
  const response = await OpenApi.Client('id')
    .addCustomHeaders({
      ['custom-header']: `dummy-value`,
    })
    .execute({
      destinationName: 'test',
      jwt: 'yourJWT'
    });
```

**Comment by tonymathew17sap** [comment]
@tomfrenken after passing JWT like you said, getting the following error:
<img width="1491" alt="Screenshot 2023-08-02 at 8 10 49 PM" src="https://github.com/SAP/cloud-sdk-js/assets/134997532/946f2a11-7e45-4837-8435-2e2a3d573283">

Do you have any idea why I'm getting this error?

**Comment by tonymathew17sap** [comment]
@tomfrenken yes the JWT is coming from approuter

**Comment by tonymathew17sap** [comment]
@tomfrenken as suggested in the fix above, I gave the iss property directly:
  `const response = await OpenApi.Client('id')
    .execute({
      destinationName: 'test',
      jwt: 'yourJWT',
     iss:'<host>/oauth/token'
    });`
    
    With this change, I'm getting 401 error:
    
<img width="647" alt="Screenshot 2023-08-03 at 5 14 51 PM" src="https://github.com/SAP/cloud-sdk-js/assets/134997532/417c498d-22bb-431c-9ccc-d76e0fd4edbc">

I guess since I have give iss directly, token exchange hasn't happened(In the request headers, I can see the same jwt token passed instead of exchange token)

**Comment by tomfrenken** [comment]
Hey @tonymathew17sap, sorry for the late response again.
We dug deeper into how the [audience validation in the xssec](https://github.wdf.sap.corp/CPSecurity/Knowledge-Base/blob/master/03_ApplicationSecurity/TokenValidation.md#3-audience-validation) library works.
One possible cause could be that your approuter is bound to an XSUAA instance that is not bound to the application, can you make sure that both use the same XSUAA instance?
If this doesn't resolve the issue, [raise the log level](https://sap.github.io/cloud-sdk/docs/js/features/logging) to debug and check if there is any more information regarding the token flow that you can share.

**Comment by tonymathew17sap** [comment]
Hi @tomfrenken debug logs:
<img width="1728" alt="Screenshot 2023-08-10 at 9 32 59 PM" src="https://github.com/SAP/cloud-sdk-js/assets/134997532/be559ec4-bc61-423d-a424-29c955bb9bbc">

Also, following this [link](https://blogs.sap.com/2022/03/07/sap-btp-security-oauth-2.0-understanding-token-exchange-2-using-destination/), I'm doing the token exchange process directly in my service after which, everything is working fine.

(First I get the token for destination service, then I use that token to make a HTTP call to the destination in the url   
`/destination-configuration/v1/destinations/<destination-name>` with request header:
`headers: {
        Authorization: 'Bearer ' + <destination service token>,
        'X-user-token': <userToken>
      }`
This http call response will have the exchanged token which I use to make the call to the target system)

**Comment by tomfrenken** [comment]
Just to reiterate, do you use the same XSUAA service instance for both your application and your approuter?
Because from the logs it looks like you use different ones which would fail the verification.
The reason they should be the same is described [in the xssec library](https://www.npmjs.com/package/@sap/xssec), which we use to verify the user JWT.

**Comment by tonymathew17sap** [comment]
@tomfrenken you are right, the issue was that there were two different XSUAA instances in the picture. I had a frontend buildapps application with authentication enabled (which used an XSUAA instance) and a backend application that used a different XSUAA instance. To confirm if two different XSUAA instances were the problem, I created an approuter and bound a common XSUAA instance to both the approuter and the backend service. After making this change, I no longer received the above error. However, I was still getting a 401 unauthenticated error from the target destination. I was getting this error because the "ForwardAuthToken" property was set in the destination, which blocked the token exchange process. After removing this property, I was able to connect to the target system without any issues! Thanks for your insight!
